Local coverimage loading

lmgonzales opened the pull request:
Allows for specification of a local file directory/directories from which to load cover images.

Resolves https://vufind.org/jira/browse/VUFIND-1137

demiankatz commented on config/vufind/config.ini:
```diff
@@ -840,6 +847,14 @@ authors         = Wikipedia
+; Note:  All coverimage LocalFile:ConfigSection must include the following keys: 
+;        filepath, filename, and imagetype.
+[LocalFile]
+filepath = /usr/local/vufind/local/coverimages
```
It's probably worth adding an explanatory comment to each of these settings.

demiankatz commented on config/vufind/config.ini:
```diff
@@ -840,6 +847,14 @@ authors         = Wikipedia
+;        filepath, filename, and imagetype.
+[LocalFile]
+filepath = /usr/local/vufind/local/coverimages
+filename = recordid
```
I think we should make the filename definition more flexible. For example, perhaps it should be a string with tokens in it, so you could do something like "%size%/%isbn%". The other advantage is that using a token-based system might allow us to extend this with some more sophisticated operations in the future... for example, I could imagine some substring operations might be helpful in scenarios where the files are split among directories to declutter the file system -- e.g. "123/456.jpg" instead of "123456.jpg" might be accessed using something like "%recordid:substr:0:3%/%recordid:substr:3%". Obviously we don't need to implement that now -- but it's nice to leave the door open to making it a future enhancement.

demiankatz replied:
It might also be helpful to make filename an optionally multi-valued setting so you could try multiple things, like:

<pre>
filename[] = "%size%/isbn/%isbn%"
filename[] = "%size%/issn/%issn%"
</pre>

demiankatz replied:
...of course, the multi-valued idea is something else we don't have to do now but could easily add in future.

lmgonzales replied:
If we go with a multivalued approach, would it make more sense to just have one multivalued field in which we specify the entire filepath (inclusive of the base directory path, filename structure and image type)? Then we just need one configuration section which allows for multiple paths to be specified in the filename[] array. What do you think?

demiankatz replied:
Yes, that makes a lot of sense... though if we go that far, do we need a config section for this at all, or should we just put the whole path after the colon in the cover setting itself, so you end up with `LocalFile:/path/to/file-pattern`? The only downside to that I can see is that it might complicate the parsing a little bit, but I don't think it would be a major obstacle.

demiankatz commented on config/vufind/config.ini:
```diff
@@ -840,6 +847,14 @@ authors         = Wikipedia
+[LocalFile]
+filepath = /usr/local/vufind/local/coverimages
+filename = recordid
+imagetype = jpg
```
Is there a strong reason to include imagetype as a separate setting rather than including it in the filename?

demiankatz commented on module/VuFind/src/VuFind/Content/Covers/LocalFile.php:
```diff
@@ -0,0 +1,150 @@
+        $imageType = strtolower(preg_replace("/(\.)(.*)/", "$0 --> $2", $imageType));
+
+        // validate that imageType is allowed
+        $allowed_imageTypes = [ "gif", "jpg", "jpeg", "png", "tif", "tiff" ];
```
Is it worth validating file type here, or should we let the higher-level code deal with that?

If we are going to do validation, it might be better to use a function that actually checks whether the contents of the file are a valid image, rather than going by the extension. I can find the details if you're not already familiar with doing this -- that way, the user isn't forced to use a particular extension, and you can avoid problems with case sensitivity in the current code here.

demiankatz commented on module/VuFind/src/VuFind/Content/Covers/LocalFile.php:
```diff
@@ -0,0 +1,150 @@
+
+        // match fileName to information in $ids array
+        if (isset($ids[$fileNameToken])) {
+            $fileName = $ids[$fileNameToken];
```
If you want to go with my token idea, you could probably get away with something like this:

<pre>
// always use size token
$tokens = ['%size%'];
$replacements = [$size];
// convert other values to tokens:
foreach ($ids as $key => $val) {
    $tokens[] = '%' . $key . '%';
    $replacements[] = $val;
}
$fileName = str_replace($tokens, $replacements, $fileNameToken);
</pre>

demiankatz replied:
...obviously this would need to get more sophisticated in future if we wanted to support functional operations on the tokens within the specification string, but we can leave that as an exercise for a future reader.

demiankatz commented on module/VuFind/src/VuFind/Content/Covers/LocalFile.php:
```diff
@@ -0,0 +1,150 @@
+            return false;
+        }
+
+        $filepath="file:" . $basePath . $fileName . "." . $imageType;
```
You should actually be prepending "file://" rather than just "file:" here to make this a valid file URL. For example, `file:/tmp/foo` is not legal, but `file:///tmp/foo` is legal. You might have to slightly adjust some of your processing code to match when you fix this.

demiankatz commented on module/VuFind/src/VuFind/Cover/Loader.php:
```diff
@@ -557,44 +569,52 @@ protected function processImageURLForSource($url, $allowCache, $source)
 
-        $image = $result->getBody();
+            // Display the image:
+            $this->contentType = 'image/jpeg';
```
If memory serves, the API code below hard-codes this value to image/jpeg because the images get converted to a JPEG as part of the downloading process. If you're going to retrieve the image files from disk without modification here, you can't assume that they are JPEGs. You should either run a conversion routine or else detect the correct MIME type. I'd recommend testing this out with a GIF, for example, to see what happens. Let me know if you need me to look up the details of any of these operations.

lmgonzales commented:
Let me see what I can do. We could parse multiple values with a : , so we'd
have LocalFile:path/to/file:path2/to/file etc...

On Thursday, June 8, 2017, Demian Katz <notifications@github.com> wrote:

> *@demiankatz* commented on this pull request.
> ------------------------------
>
> In config/vufind/config.ini
> <https://github.com/vufind-org/vufind/pull/984#discussion_r120912436>:
>
> > @@ -840,6 +847,14 @@ authors         = Wikipedia
>  url = "https://api.vlb.de/api/v1/cover/"
>  ; token = "XXXXXX-XXXX-XXXXX-XXXXXXXXXXXX"
>
> +; This section is needed for LocalFile cover loading.
> +; Note:  All coverimage LocalFile:ConfigSection must include the following keys:
> +;        filepath, filename, and imagetype.
> +[LocalFile]
> +filepath = /usr/local/vufind/local/coverimages
> +filename = recordid
>
> Yes, that makes a lot of sense... though if we go that far, do we need a
> config section for this at all, or should we just put the whole path after
> the colon in the cover setting itself, so you end up with
> LocalFile:/path/to/file-pattern? The only downside to that I can see is
> that it might complicate the parsing a little bit, but I don't think it
> would be a major obstacle.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/vufind-org/vufind/pull/984#discussion_r120912436>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMc0UbVIarCJ3xNmLKm0VvGreRq8Y6Syks5sCAyXgaJpZM4NzaL1>
> .
>

demiankatz commented:
It might be safer to go with `LocalFile:/path/to/file,LocalFile:path2/to/file` instead of using a colon-separated list on a single call to LocalFile, since under Windows there could be a `c:\` in the path which could cause confusion.

lmgonzales commented:
@demiankatz, 
Thanks for the suggestions for improving the code. I've re-written/updated the following in the latest commit: 

**config.ini** - refactored and added explanatory notes about LocalFile:PathToFile configuration. This now allows for filepath and image type tokens as well as declaration of multiple paths. I figure we can add in code in the future to cover the scenarios where the files are split among directories -- e.g.  your example of "%recordid:substr:0:3%/%recordid:substr:3%"

**LocalFile.php** - major re-write to handle tokens (thank you for the example code!), validate image types and mime types, and fix the "file://" prepending of the filePath variable. 
 
**Cover/Loader.php** - updated to check for "file://" in $url, and also updated the $this->contentType to retrieve the mime type. Since we already validated the file in the LocalFile.php script, it's fine to just pull the mime type here without checking it again. 

Please let me know if you see anything else you would recommend for improvement. I think I have addressed all of your comments/suggestions, but if I missed any, please let me know.

demiankatz commented:
@lmgonzales, thanks again for all of the work on this! It's working for me. A few comments:

1.) I was able to delete some code and configurations that we didn't need any more based on the way things have evolved.

2.) It might be more clear to treat the VuFind base path / local path as the same kinds of tokens as everything else. Having some $TOKEN values and some %token% values may be unnecessarily confusing. Also, note that you don't have to use getenv to get the environment values, since they'll already be set for you as constants. I would suggest simply initializing the $tokens and $replacements arrays like this, and removing the special lines dealing with environment variables:

<pre>
$tokens = ['%vufind-home%', '%vufind-local-dir%'];
$replacements = [APPLICATION_PATH, LOCAL_OVERRIDE_DIR];
</pre>

3.) I'm not sure that it's necessary to be quite so concerned about file extensions matching MIME types; I'd be satisfied if we just checked that the MIME type was good. Basically just create an $allowedMimeTypes array, and then check that the type is in the array. I wouldn't bother with having all of the extension-specific tokens like %Jpg%, etc.; instead, I'd just make the extension part of the file path pattern. The only extension-related token I would keep is %anyimage%, which you could process with a special case. I think if you rip some of this complexity out, the code will naturally shrink and become more clear, but you won't really lose anything in the way of security or functionality.

Let me know what you think! I'm happy to help with implementation of any or all of these changes next week if you'd like me to. Probably won't take terribly long, but I'm out of time for today. Have a good weekend!

demiankatz commented:
Just noticed a significant bug (that's my fault for not realizing earlier) -- the 'isbn' key of the $ids array is not a string but an object with get10() and get13() methods for retrieving the different possible ISBN formats. If you send an ISBN to the LocalFile handler now, it causes a fatal string conversion error. We'll probably want to implement %isbn10% and %isbn13% tokens and convert these to the appropriate method calls. The current loop that sets up default tokens should include an is_string() check so it can skip unsupported non-string values (for future safety, in case other objects are introduced later).

lmgonzales commented:
Thanks Demian!



1)      Passing in config to LocalFile.php:  Okay sounds good. We can
always add the config/factory code back in at a later date if we need it.

2)      Environment variables: Great idea. I’ll get this updated.

3)      Mime types: Sounds good. I’ll get these modifications taken care of
today.



-Leila



*From:* Demian Katz [mailto:notifications@github.com]
*Sent:* Friday, June 09, 2017 8:29 AM
*To:* vufind-org/vufind
*Cc:* Leila Gonzales; Mention
*Subject:* Re: [vufind-org/vufind] Local coverimage loading (#984)



@lmgonzales <https://github.com/lmgonzales>, thanks again for all of the
work on this! It's working for me. A few comments:

1.) I was able to delete some code and configurations that we didn't need
any more based on the way things have evolved.

2.) It might be more clear to treat the VuFind base path / local path as
the same kinds of tokens as everything else. Having some $TOKEN values and
some %token% values may be unnecessarily confusing. Also, note that you
don't have to use getenv to get the environment values, since they'll
already be set for you as constants. I would suggest simply initializing
the $tokens and $replacements arrays like this, and removing the special
lines dealing with environment variables:

$tokens = ['%vufind-home%', '%vufind-local-dir%'];

$replacements = [APPLICATION_PATH, LOCAL_OVERRIDE_DIR];

3.) I'm not sure that it's necessary to be quite so concerned about file
extensions matching MIME types; I'd be satisfied if we just checked that
the MIME type was good. Basically just create an $allowedMimeTypes array,
and then check that the type is in the array. I wouldn't bother with having
all of the extension-specific tokens like %Jpg%, etc.; instead, I'd just
make the extension part of the file path pattern. The only
extension-related token I would keep is %anyimage%, which you could process
with a special case. I think if you rip some of this complexity out, the
code will naturally shrink and become more clear, but you won't really lose
anything in the way of security or functionality.

Let me know what you think! I'm happy to help with implementation of any or
all of these changes next week if you'd like me to. Probably won't take
terribly long, but I'm out of time for today. Have a good weekend!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/vufind-org/vufind/pull/984#issuecomment-307420803>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AMc0UZ0xg4qsnjmPLu8Qk6rLs37rc2bLks5sCWTSgaJpZM4NzaL1>
.

lmgonzales commented:
Ok. I was wondering about that. I’ll go grab the get10() and get13() code
and process the isbn tokens accordingly. Thanks for catching that!



*From:* Demian Katz [mailto:notifications@github.com]
*Sent:* Friday, June 09, 2017 8:35 AM
*To:* vufind-org/vufind
*Cc:* Leila Gonzales; Mention
*Subject:* Re: [vufind-org/vufind] Local coverimage loading (#984)



Just noticed a significant bug (that's my fault for not realizing earlier)
-- the 'isbn' key of the $ids array is not a string but an object with
get10() and get13() methods for retrieving the different possible ISBN
formats. If you send an ISBN to the LocalFile handler now, it causes a
fatal string conversion error. We'll probably want to implement %isbn10%
and %isbn13% tokens and convert these to the appropriate method calls. The
current loop that sets up default tokens should include an is_string()
check so it can skip unsupported non-string values (for future safety, in
case other objects are introduced later).

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/vufind-org/vufind/pull/984#issuecomment-307422393>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AMc0Uez9mbkz_-1kfcpPl64oMqpGrt4Zks5sCWYfgaJpZM4NzaL1>
.[image: Image removed by sender.]

demiankatz commented:
Great, thanks! Let me know if you have any questions or problems with the ISBN stuff... should just be a matter of doing:

<pre>
$tokens[] = '%isbn10%';
$replacements[] = $ids['isbn']->get10();
</pre>

(or something along those lines).

lmgonzales commented:
@demiankatz, 
1) I've updated the config.ini file so that it takes base path/ local path tokens instead of environment variables, and so it takes %isbn10% and %isbn13% tokens rather than %isbn%.
2) I've refactored the LocalFile.php code so that it addresses your comments above, namely:
   a) only checking for mime types (instead of matching mime type to file extension)
   b) making sure the $ids tokens are strings before they are added to the tokens[] and replacements[] arrays.
  c) dealing with %isbn10% and %isbn13% tokens
  d) general refactoring to make the code more concise.

demiankatz commented:
@lmgonzales, I've put in a little more work on this. Noteworthy changes:

- I've condensed the code a little more (sorry, I can't help myself)
- I've added some missing logic for creating cache paths for records that only have a 'recordid' value in the ID array.
- I've made %anyimage% a bit less case-sensitive (though it's not 100% perfect in that regard).
- I've moved some things to class properties in the LocalFile class, which should make it easier to add additional file extensions or MIME types through subclassing.
- I've greatly simplified the file type handling -- there is no longer file extension validation, since the MIME type logic ensures valid content. Now you don't need to use tokens like %jpg%, you can just include ".jpg" as part of your path. However, if you drop in the %anyimage% token, you still get wildcard matching of known image extensions.

This last change (the %anyimage% stuff) is perhaps the most likely to be controversial, since it does significantly change the logic. I think it's a change for the better, but I'm happy to discuss it if you disagree.

Anyway, as far as I can tell, this is ready to merge now -- but I'll wait for you to give it a review and re-test before I push it through. Please let me know what you think!

Thanks again!

lmgonzales commented:
@demiankatz ,
Thank you very much for the improvements and streamlining of the code. It looks much better. The %anyimge% handling changes you've made make sense to me. Thanks for implementing it!

demiankatz commented:
@lmgonzales, thanks for the double-check and for all of your work on this. I've gone ahead and merged it to master, after adding support for a %source% parameter (on reflection, it seemed incomplete to support %recordid% without %source%, and it only took two minutes to add).

lmgonzales commented:
@demiankatz , thanks for adding in the support for the %source% token, and for merging this PR into master.

lmgonzales commented:
@demiankatz, 
Just had a quick question for you regarding the Cover/Loader.php file in the determineLocalFile() method. Do we need a separate case for where %recordid% is set but %source% is not?

demiankatz commented:
@lmgonzales, at this point, no -- you can't have one without the other as currently implemented. In theory we could add logic to say that if source is unset, it should default to the default backend constant... but at the moment I don't think there's a need to worry about that.

lmgonzales commented:
@demiankatz - okay. Thanks for the clarification!
